databricks/koalas/frame.py

```diff
@@ -9122,6 +9122,179 @@ def gen_new_column_labels_entry(column_labels_entry):
+                       num_legs  num_arms
+        type   name
+        mammal dog            4         0
+               cat            4         0
+               monkey         2         2
+
+        >>> df.rename_axis(index={'type': 'class'}).sort_index()  # doctest: +NORMALIZE_WHITESPACE
+                       num_legs  num_arms
+        class  name
+        mammal cat            4         0
+               dog            4         0
+               monkey         2         2
+
+        >>> df.rename_axis(index=str.upper).sort_index()  # doctest: +NORMALIZE_WHITESPACE
+                       num_legs  num_arms
+        TYPE   NAME
+        mammal cat            4         0
+               dog            4         0
+               monkey         2         2
+        """
+
+        def gen_names(v, curnames):
+            if is_scalar(v):
+                newnames = [v]
+            elif is_list_like(v) and not is_dict_like(v):
+                newnames = list(v)
+            elif is_dict_like(v):
+                newnames = [v[name] if name in v else name for name in curnames]
+            elif callable(v):
+                newnames = [v(name) for name in curnames]
+            else:
+                raise ValueError(
+                    "`mapper` or `index` or `columns` should be "
+                    "either dict-like or function type."
+                )
+
+            if len(newnames) != len(curnames):
+                raise ValueError(
+                    "Length of new names must be {}, got {}".format(len(curnames), len(newnames))
+                )
+
+            return [
+                name if name is None or isinstance(name, tuple) else (name,) for name in newnames
+            ]
+
+        if mapper is not None and (index is not None or columns is not None):
+            raise TypeError("Cannot specify both 'mapper' and any of 'index' or 'columns'.")
+
+        if mapper is not None:
+            axis = validate_axis(axis)
+            if axis == 0:
+                index = mapper
+            elif axis == 1:
+                columns = mapper
+
+        column_label_names = (
+            gen_names(columns, self.columns.names)
+            if columns is not None
+            else self._internal.column_label_names
+        )
+        index_names = (
+            gen_names(index, self.index.names) if index is not None else self._internal.index_names
+        )
+        index_map = OrderedDict(zip(self._internal.index_spark_column_names, index_names))
+
+        spark_frame = self._internal.resolved_copy.spark_frame
+        internal = InternalFrame(
+            spark_frame=spark_frame,
+            index_map=index_map,
+            column_labels=self._internal.column_labels,
+            data_spark_columns=[
+                scol_for(spark_frame, col) for col in self._internal.data_spark_column_names
+            ],
+            column_label_names=column_label_names,
+        )
```

Comment:
Oh I think we don't need to create a new `InternalFrame` and maybe can just simply copy the `index_map` and `column_label_names` here ??

Because this API updates only the name of index or columns rather than any other features.

```python
        internal = self._internal.copy(
            index_map=index_map,
            column_label_names=column_label_names,
        )
```

Reply:
Definitely !